### PR TITLE
1.19 launchpad and elevator, textures by Spearkiller

### PR DESCRIPTION
**Pull request by Ayutac**
Launchpad and Elevator for TR. Whats missing is the mixin for the player jumping/sneaking to activate it, that should call teleportUp/Down on the ElevatorBlockEntity.

(Note that the Launchpad is also contained in this PR rn)

### CODE REVIEW

File: src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+import techreborn.init.TRBlockEntities;
+import techreborn.init.TRContent;
+
+import java.util.List;
+import java.util.Optional;
+
+public class ElevatorBlockEntity extends PowerAcceptorBlockEntity implements IToolDrop, BuiltScreenHandlerProvider {
+
+	public static final int MIN_HEIGHT = -64;
+	public static final int MAX_HEIGHT = 319;
```
Comment (modmuss50) [inline src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java]: You should use `world.isOutOfHeightLimit()` or `getBottomY()` / `getTopY()` as diffrent worlds can have different heights

File: src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+	public ElevatorBlockEntity(BlockPos pos, BlockState state) {
+		super(TRBlockEntities.ELEVATOR, pos, state);
+	}
+
+	public static boolean isRunning(final World world, final BlockPos pos) {
+		final BlockEntity entity = world.getBlockEntity(pos);
+		if (!(entity instanceof ElevatorBlockEntity)) {
+			return false;
+		}
```
Comment (modmuss50) [inline src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java]: Dont make this a static method, removes the need for a 2nd getBlockEntity() call.

File: src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+		return ((ElevatorBlockEntity)entity).getStored() > 0;
+	}
+
+	public static boolean isFree(final World world, final BlockPos pos) {
```
Comment (modmuss50) [inline src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java]: No need for this to be static.

File: src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+		return world.getBlockState(pos.up()).isAir() && world.getBlockState(pos.up().up()).isAir();
+	}
+
+	public static boolean isValidTarget(final World world, final BlockPos pos) {
```
Comment (modmuss50) [inline src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java]: No need for this to be static. Et al.

File: src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+		return Math.min(Math.abs(endPos.getY()-startPos.getY())*TechRebornConfig.elevatorEnergyPerBlock,0);
+	}
+
+	protected boolean teleport(World world, BlockPos pos, PlayerEntity player, BlockPos targetPos) {
+		final int energy = energyCost(pos, targetPos);
+		if (getStored() < energy) {
+			return false;
+		}
+		world.playSound(null, pos, SoundEvents.ENTITY_ENDERMAN_TELEPORT, SoundCategory.BLOCKS, 1f, 1f);
+		player.teleport(targetPos.getX(), targetPos.getY(), targetPos.getZ());
```
Comment (modmuss50) [inline src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java]: Use FabricDimensions.teleport, its much better.

File: src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+			return Optional.of(downPos);
+		}
+		return Optional.empty();
+	}
+
+	public static int energyCost(BlockPos startPos, BlockPos endPos) {
+		return Math.min(Math.abs(endPos.getY()-startPos.getY())*TechRebornConfig.elevatorEnergyPerBlock,0);
+	}
+
+	protected boolean teleport(World world, BlockPos pos, PlayerEntity player, BlockPos targetPos) {
```
Comment (modmuss50) [inline src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java]: No need for the World world param, you can get it from the BE.

File: src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+	// PowerAcceptorBlockEntity
+	@Override
+	public void tick(World world, BlockPos pos, BlockState state, MachineBaseBlockEntity blockEntity) {
+		super.tick(world, pos, state, blockEntity);
+		if (world == null || getStored() <= 0 || !isActive(RedstoneConfiguration.POWER_IO)) {
+			return;
+		}
+
+		Optional<BlockPos> upTarget = null;
```
Comment (modmuss50) [inline src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java]: I dont think you need both of these, you can just have target. Will let you simplify the code a bit.

**Review by modmuss50 — CHANGES_REQUESTED** [review]

File: src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java (hunk 2 above)
Comment (Ayutac) [reply src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java]: nope, this is to check if another elevator is running when calculating the target from the current elevator

File: src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java (hunk 7 above)
Comment (Ayutac) [reply src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java]: the idea was when mutliple players teleport at the same time, to only calculate their target once. still holds for sneaking, has been made redundant for jumping

File: src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java (hunk 1 above)
Comment (Ayutac) [reply src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java]: I was looking for those, thanks!

File: src/main/java/techreborn/packets/ServerboundPackets.java [hunk 8]
```diff
@@ -171,6 +174,29 @@ public static void init() {
...
+
+			server.execute(() -> {
+				BlockEntity blockEntity = player.world.getBlockEntity(pos);
+				if (blockEntity instanceof LaunchpadBlockEntity) {
+					((LaunchpadBlockEntity) blockEntity).handleGuiInputFromClient(buttonAmount);
+				}
+			});
+		}));
+
+		NetworkManager.registerServerBoundHandler(new Identifier("reborncore", "jump"), (server, player, handler, buf, responseSender) -> {
```
Comment (modmuss50) [inline src/main/java/techreborn/packets/ServerboundPackets.java]: You shouldnt listen to other mods packets like this. The packet should be sent from TR and recieved by TR

File: src/main/java/techreborn/packets/ServerboundPackets.java [hunk 9]
```diff
@@ -171,6 +174,29 @@ public static void init() {
...
+			BlockPos pos = buf.readBlockPos();
+
+			server.execute(() -> {
+				BlockEntity blockEntity = player.getWorld().getBlockEntity(pos.down());
+				if (blockEntity instanceof ElevatorBlockEntity elevator) {
+					elevator.teleportUp(player.getWorld(), pos.down(), player);
```
Comment (modmuss50) [inline src/main/java/techreborn/packets/ServerboundPackets.java]: You should check that the player is within a few blocks of the evevator, as right now someone could easily send this packet from anywhere in the world and get teleported.

**Review by modmuss50 — CHANGES_REQUESTED** [review]

File: src/main/java/techreborn/packets/ServerboundPackets.java (hunk 8 above)
Comment (Ayutac) [reply src/main/java/techreborn/packets/ServerboundPackets.java]: But the sending happens in the mixin? How am i supposed to have the mixin in core and the sending in TR?

File: src/main/java/techreborn/packets/ServerboundPackets.java (hunk 9 above)
Comment (Ayutac) [reply src/main/java/techreborn/packets/ServerboundPackets.java]: doesnt the code require that the block below the player is an elevator?

File: src/main/java/techreborn/packets/ServerboundPackets.java (hunk 8 above)
Comment (modmuss50) [reply src/main/java/techreborn/packets/ServerboundPackets.java]: You should create an event in RC, RC should not be sending jump packets.

File: src/main/java/techreborn/packets/ServerboundPackets.java (hunk 9 above)
Comment (modmuss50) [reply src/main/java/techreborn/packets/ServerboundPackets.java]: Doesnt look like the code does. You cannot trust any packet data coming from the client.

File: src/main/java/techreborn/packets/ServerboundPackets.java (hunk 8 above)
Comment (Ayutac) [reply src/main/java/techreborn/packets/ServerboundPackets.java]: I'm out of my depth at this point then.

File: src/main/java/techreborn/packets/ServerboundPackets.java (hunk 9 above)
Comment (Ayutac) [reply src/main/java/techreborn/packets/ServerboundPackets.java]: Ah, I see my mistake now. But I would wait with fixing until the "event" is properly implemented

**Review by modmuss50 — APPROVED** [review]
Awesome 👍

**Pull request merged**